### DISCUSSION TITLE
Comparison to similar projects?

### DISCUSSION THREAD

**Discussion by pietz** [pelme/htpy#7] [General]
Can you share some thoughts on htpy while comparing it with current alternatives? It's not clear to me what it improves over libraries like dominate or lxml.builder.

Thanks

**Comment by pelme** [comment]
It is very similar. htpy has some helpers that makes writing html a bit more convenienes like automatic doctype, “#id.class” short hand and built in type hints. See the docs for more information on the usages. 

I have looked at all similar projects I could find and tried to combine the best ideas. I don’t really think htpy brings anything new to the table apart from bringing old ideas into modern python 🙂 there is a list of related projects at https://htpy.dev/references/. I will add lxml.builder there too!

**Comment by pietz** [comment]
Thanks! Have you run some performance tests against others?

**Reply by pelme** [reply]
We have compared to Django and Jinja templates: https://htpy.dev/faq/#how-does-htpy-performance-compare-to-django-or-jinja-templates

Given that it has been fast enough for us, we have not done further benchmarks. We have been floating the idea of building the module in rust or cython but it is not pressing to us.

**Reply by pietz** [reply]
That makes sense. Thanks for getting back to me so quickly.

**Reply by pelme** [reply]
You're welcome! 🙂

**Reply by pietz** [reply]
I extended your benchmark to dominate and fast_html, and htpy was quite a bit faster than all of them except for jinja2 of course.

```
htpy: 0.16253904199402314 seconds - /var/folders/rm/y4mfyltx67d84r0n1h3ccgp80000gn/T/tmpqe9ecfq_/htpy_table.html
django: 0.3162372079968918 seconds - /var/folders/rm/y4mfyltx67d84r0n1h3ccgp80000gn/T/tmpqe9ecfq_/django_table.html
jinja2: 0.008120957994833589 seconds - /var/folders/rm/y4mfyltx67d84r0n1h3ccgp80000gn/T/tmpqe9ecfq_/jinja2_table.html
dominate: 0.6329318329953821 seconds - /var/folders/rm/y4mfyltx67d84r0n1h3ccgp80000gn/T/tmpqe9ecfq_/dominate_table.html
fast_html: 0.5624827080027899 seconds - /var/folders/rm/y4mfyltx67d84r0n1h3ccgp80000gn/T/tmpqe9ecfq_/fast_html_table.html
```

More importantly though, you're right, benchmarking this is kind of unnecessary because all of them are fast enough for real world scenarios. I see your point.

Anyway, I feel like you could share a few words about design and syntax philosophy maybe through a blog post or in the docs. I have spotted a couple of things I prefer over alternatives but most of them weren't clear to me initially. You probably know what I mean. I would be very interested in getting a opinionated intro to htpy.

**Reply by pelme** [reply]
I deliberativ did not put information about performance on the front page. Many frameworks/libraries do that but that is not where I want to put the focus. It is nice to see that htpy is not slower than some similar libraries though, thanks for sharing. 🙂 

I am not sure what to put in post like that. Have tried to show (without too many opinions though) useful syntax I could think of at https://htpy.dev/usage/ and also the patterns we use at https://htpy.dev/common-patterns/. Is there some particular info that you miss from the docs? All feedback on the docs are very welcome!

**Reply by davepeck** [reply]
Many moons later, I have some new anecdata on perf:

I built a new performance test, rendering the most complex page I've got in the site I'm currently building. The page is an adaptation of a full Tailwind UI (commercial) template, so I can't share the code publicly alas. But: it has a lot of markup and requires many sub-functions to render.

I used LLMs to help me build a Django/Jinja2 template equivalent. (They are... not something I want to maintain.) 

My perf test runs for each of the three (htpy, Django, jinja2) 10k times and averages the results. It also configures Django to use its cache loader. 

The results were:

- **`htpy` and Django were tied**. Technically,  Django edged out `htpy` by 0.00002 seconds 😂 &mdash; aka well below my faith in the accuracy of my numbers
- Jinja2 was about 3.5x faster, no surprise.

PS: converting an intended-for-React template with complex transitions, etc. into htpy, htmx, and a sprinkling of entirely optional javascript is... very instructive.

**Reply by pelme** [reply]
Very cool, thanks for the update! It sounds like a much better benchmark than `scripts/benchmark_big_table.py`. Is there any way you could replace the commercial Tailwind UI classes with something boilerplate and share it (without too much work)?

I run `scripts/benchmark_big_table.py` regularly when making changes to make sure I do not completely screw up the performance but it is a very lacking benchmark.
